File: vscode/src/utils/inlay/GithubInlayUtils.ts
import * as vscode from "vscode";
import { I18nKey } from "../../common/I18nKey";
import { Result } from "../../common/Result";
import { GithubRepoInfoService } from "../service/GithubRepoInfoService";
import { RepoKey } from "../service/RepoKey";
import { I18n } from "../I18n";
import { ProgressUtils } from "../ProgressUtils";
import { UiUtils } from "../UiUtils";

export class GithubInlayUtils {
  static addRepoInlay(
    hints: vscode.InlayHint[],
    position: vscode.Position,
    repoKey: RepoKey,
  ): void {
    const { owner, repo } = repoKey;
    const repoId = `${owner}/${repo}`;
    const res = GithubRepoInfoService.getInstance().getRepoInfo(owner, repo);

    let label = "";
    const retryToken = `github:${owner}/${repo}`;
    const githubUrl = `https://github.com/${owner}/${repo}`;

    switch (res.result) {
      case Result.NONE:
        label = I18n.message(I18nKey.loadingGithub);
        void Promise.resolve(
          ProgressUtils.runBackground(
            `DepLens: Fetch GitHub ${owner}/${repo}`,
            () => GithubRepoInfoService.getInstance().fetchRepoInfo(owner, repo),
          ),
        ).catch((error: unknown) => {
          console.warn(`Failed to load repo info for ${repoId}`, error);
        });
        break;

      case Result.PENDING:
        label = I18n.message(I18nKey.loadingGithub);
        break;

      case Result.SUCCESS: {
        const stars = res.data?.stars ?? "0";
        const updated = res.data?.updatedDate ?? "N/A";
        label = `⭐ ${stars} • ${I18n.message(I18nKey.lastUpdated)} ${updated}`;
        break;
      }

      case Result.FAILURE:
        label = I18n.message(I18nKey.failedGithub);
        break;
    }

    UiUtils.addInlay(
      hints,
      position,
      label,
      label,
      githubUrl,
      retryToken
    );
  }
}

File: zed/lsp/src/utils/inlay/github_inlay_utils.rs
use crate::common::i18n_key;
use crate::common::result::Result;
use crate::lsp::{InlayHint, Position};
use crate::utils::i18n::I18n;
use crate::utils::service::github_repo_info_service::GithubRepoInfoService;
use crate::utils::ui_utils::UiUtils;

pub struct GithubInlayUtils;

impl GithubInlayUtils {
    pub fn add_repo_inlay(
        hints: &mut Vec<InlayHint>,
        position: Position,
        owner: String,
        repo: String,
    ) {
        let repo_id = format!("{owner}/{repo}");
        let res = GithubRepoInfoService::get_instance().get_repo_info(&owner, &repo);

        let retry_token = format!("github:{owner}/{repo}");
        let github_url = format!("https://github.com/{owner}/{repo}");

        let label = match res.result {
            Result::None => {
                GithubRepoInfoService::get_instance().start_fetch_repo_info(owner.clone(), repo.clone(), repo_id);
                I18n::message(i18n_key::LOADING_GITHUB)
            }
            Result::Pending => I18n::message(i18n_key::LOADING_GITHUB),
            Result::Success => {
                let data = res.data.expect("success result has data");
                format!(
                    "⭐ {} • {} {}",
                    data.stars,
                    I18n::message(i18n_key::LAST_UPDATED),
                    data.updated_date
                )
            }
            Result::Failure => I18n::message(i18n_key::FAILED_GITHUB),
        };

        UiUtils::add_inlay(
            hints,
            position,
            label.clone(),
            label,
            Some(github_url),
            Some(retry_token),
        );
    }
}
